### ISSUE TITLE
support changing rounding modes?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#11970]
As a Chapel user, I'd like to be able to write programs that operate with different floating point rounding modes. Changing the rounding mode will allow me to implement interval arithmetic or otherwise compute the error of my computation.

C offers the ability to query and change rounding modes with the `fegetround()` and `fesetround()` calls. Should Chapel support something along these lines?

@dmk42 pointed out:

> If we decide to support changing rounding modes, then each change creates a barrier beyond which floating-point code cannot be moved. The reason is that most floating-point units serialize on the mode change. I've heard that there are machines that encode the rounding mode into the instruction instead of setting some state, but I haven't actually seen one. In C, the rounding mode is manipulated by calling fegetround() and fesetround(), which require the following pragma to b

**Comment by dmk42** [comment]
This is a long-term issue because in the near term, LLVM may hoist arithmetic up above a call to `fesetround()`.  That is being worked on from the LLVM side.

**Comment by bradcray** [comment]
In ancient history, we had users who thought that Fortran (eventually) did a really good job with providing knobs and introspection around IEEE floating point and suggested we copy what they did.  I've always intended to make good on that suggestion by learning what Fortran has, but it's never made it to the top of the list.

**Comment by damianmoz** [comment]
Do you mean this sort of stuff
```
https://wg5-fortran.org/N1351-N1400/N1378.pdf
```
You might also look at some of the ideas of Nelson Beebe
```
w.math.utah.edu/~beebe/software/ieee
```
And there was Apple's SANE environment.

**Comment by mppf** [comment]
I'm trying to have more issues on specific topics, so I'd like this one to continue to be about changing rounding modes. I created other issues for what I saw as the future work / design questions around IEEE exception handling.

Additionally I'm not sure what other "knobs and introspection" could be interesting in this context. The only other thing I can think of is choosing the allowable error bound for optimizations and/or library functions.

**Comment by damianmoz** [comment]
I agree, as you say, I think this topic should stick to rounding modes.

Maybe @bradcray will elaborate on what "knows and introspection" he meant and if they fall outside the rounding mode topic, you can start a new one. I thought he was talking about the whole Fortran interface.

The error bounds for the elementary functions are well predefined as having an upper bound 1*ULP (unit in the last place). That said, Intel seems to product fast, less accurate versions of some routines. How you pre-select one or the other I would have thought was done at linkage time. That is a big job if you want to provide such a library within the Chapel context and do it both comprehensively and consistently. A serious research project and one which has implications as to the underlying architecture.

I will put together a list of primitives which need to be available to work on IEEE 754 numbers. But not today. By primitive, I mean operations which are done in line by the compiler although I can provide a
C version of what I mean as functions. I use these with C/C++ a inline functions a they have unacceptable overheads with calls. These rely on being able to work on a 64-bit real as an unsigned 64-bit integer without a function call over, a programming technique which needs to be done with extreme care and preferably almost never.

**Comment by damianmoz** [comment]
See the following for some research into faster and less accurate elementary functions:
```
https://arxiv.org/pdf/1811.10274.pdf
```
You probably have loads of information on the Intel approximate stuff but it probably tells you how to use it properly rather than what makes it tick. The latter information would, I assume, be more valuable coming from your perspective.
